csharp/src/Ice/Reference.cs

```diff
@@ -658,33 +757,68 @@ static void StartQueryOption(StringBuilder sb, ref bool firstOption)
                 IInvocationObserver? observer = ObserverHelper.GetInvocationObserver(proxy,
                                                                                      request.Operation,
                                                                                      request.Context);
-                int attempt = 1;
+
                 // If the request size is greater than Ice.RetryRequestSizeMax or the size of the request
                 // would increase the buffer retry size beyond Ice.RetryBufferSizeMax we release the request
                 // after it was sent to avoid holding too much memory and we wont retry in case of a failure.
                 int requestSize = request.Size;
                 bool releaseRequestAfterSent =
                     requestSize > reference.Communicator.RetryRequestSizeMax ||
                     !reference.Communicator.IncRetryBufferSize(requestSize);
+
                 try
                 {
+                    Connection? connection;
+                    bool cached;
+                    IReadOnlyList<Endpoint> endpoints;
+                    List<Connector>? connectors;
+                    OutgoingConnectionFactory connectionFactory = reference.Communicator.OutgoingConnectionFactory;
+
+                    (connection, cached, endpoints, connectors) =
+                        await reference.GetExistingConnectionAsync(cancel).ConfigureAwait(false);
+
+                    int nextConnector = 0;
+                    int retryAttempt = 0;
+                    bool triedAllConnectors = false;
+                    List<Connector>? excludedConnectors = null;
+                    int maxRetryAttempts = reference.Communicator.RetryMaxAttempts;
```

Comment:
maxRetryAttempts does not mean max attempts.

Reply:
renamed to `Communicator.MaxAttempts`